docs/handler-type-comparison.md

```diff
@@ -0,0 +1,131 @@
+
+If you're new to Dogma, start by familiarising yourself with the key [concepts].
+
+> [!TIP]
+> Consult the [glossary] for definitions of terms used in this document.
+
+## Comparison matrix
+
+|                        | Aggregate | Process  | Integration |  Projection  |
+| ---------------------- | :-------: | :------: | :---------: | :----------: |
+| Command messages       |     ↓     |    ↑     |      ↓      |              |
+| Event messages         |     ↑     |    ↓     |      ↑      |      ↓       |
+| Timeout messages       |           |    ⇅     |             |              |
+| [State]                |     ✓     |    ✓     |             |              |
+| [External data access] |           |    ~     |      ✓      |      ✓       |
+| [Isolation boundary]   | instance  | instance |   command   | event-stream |
+| [Event stream binding] | instance  |    —     |   command   |      —       |
+
+#### Legend
+
+- `✓` supported
+- `~` supported, but not recommended
+- `↓` consumes
+- `↑` produces
+- `⇅` produces & consumes
+- `—` not applicable
+
+## Definitions
+
+### State
+
+Aggregates and processes maintain state across many separate instances.
+The engine routes each message to a specific instance using an identifier
+provided by the application.
+
+Each aggregate instance records events, and each process instance holds data.
+Together, this represents the application's state.
+
+Although both integration and projection message handlers may use or store data,
+it's not considered state within Dogma. See the [definition of state] in the
+glossary for further clarification.
+
+### External data access
+
+External data is any information not managed by the engine — including
+projection data, third-party APIs, the system clock, and so on.
+
+- Aggregate message handlers must not read or write external data. All logic
+  must rely solely on information within command messages and the instance's
+  state. This restriction ensures that aggregates are deterministic — they
+  produce the same events for the same sequence of commands.
+
+- Process message handlers may read external data during message routing and
+  handling, primarily to access the application's projection data, though this
+  is not recommended. Wherever possible, all logic should rely solely on
+  information within event messages and the instance's state. This restriction
+  ensures that processes are deterministic — they produce the same behavior for
+  the same sequence of events.
+
+- Integration message handlers may read and write external data — they exist
+  specifically to interact with external systems.
+
+- Projection message handlers may read and write external data. Where possible,
+  all logic should rely solely on information within event messages and the
+  existing projection data. This restriction ensures that projections are
+  self-contained and deterministic — they produce the same data for the same
+  sequence of events.
```

Comment:
This mentions external data use being a "restriction" yet it is allowed and not mentioned as "not recommended".

Reply:
"restriction" is referring to the prior sentence:

> Where possible, all logic should rely solely on information within event messages and the existing projection data.